### PR TITLE
Remember me

### PR DESCRIPTION
**Pull request by mrcuong84**

### CODE REVIEW

File: spec/controllers/sessions_controller_spec.rb [hunk 1]
```diff
@@ -7,16 +7,50 @@
...
+      end
+
+      it 'render to new' do
+        expect(subject).to render_template(:new)
+      end
+    end
+
+    context 'valid param input not checked remember me' do
+      subject do
+        post :create, params: {
```
Comment (banguyenvareal) [inline spec/controllers/sessions_controller_spec.rb:26]: `should use shared_context`

Comment (mrcuong84) [reply spec/controllers/sessions_controller_spec.rb:26]: already fix

**Review by banguyenvareal — APPROVED** [review]

File: app/models/user.rb [hunk 2]
```diff
@@ -16,4 +18,27 @@ class User < ApplicationRecord
...
+
+  def create_user_digest(string)
+    BCrypt::Password.create(string, cost: BCrypt::Engine::DEFAULT_COST)
+  end
+
+  def create_user_new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def create_token_and_update_remember_digest
```
Comment (hieunguyentrung) [inline app/models/user.rb]: `generate_remember_token!`

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -1,9 +1,39 @@
...
+  end
+
+  def find_and_assign_by_cookies
+    user = User.find_by(id: cookies.signed[:user_id])
+    return unless user&.authenticated?(cookies[:remember_token])
+
+    @current_user = user
+  end
+
+  def remember_and_assign_cookies(user)
```
Comment (hieunguyentrung) [inline app/helpers/sessions_helper.rb]: `remember_user`

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -1,9 +1,39 @@
...
+    user.create_token_and_update_remember_digest
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def forget_and_delete_cookies(user)
```
Comment (hieunguyentrung) [inline app/helpers/sessions_helper.rb]: `forget_user`

File: app/models/user.rb [hunk 5]
```diff
@@ -16,4 +18,27 @@ class User < ApplicationRecord
...
+    update_attribute(:remember_digest, create_user_digest(remember_token))
+  end
+
+  def authenticated?(remember_token)
+    return false if remember_digest.nil?
+
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  def remove_remember_digest_value
```
Comment (hieunguyentrung) [inline app/models/user.rb]: `remove_remember_digest`

File: app/models/user.rb [hunk 6]
```diff
@@ -16,4 +18,27 @@ class User < ApplicationRecord
...
+    update_attribute(:remember_digest, nil)
```
Comment (hieunguyentrung) [inline app/models/user.rb:42]: `update_attribute(remember_digest: nil)`

File: spec/controllers/sessions_controller_spec.rb [hunk 7]
```diff
@@ -1,22 +1,44 @@
...
-  let(:password) { 'Aa123456' }
-  let(:other_password) { 'Aa00000' }
-  let(:user) do
-    create(:user, password: password, password_confirmation: password)
-  end
-
+  include_context 'user_login_params'
   describe '#create' do
-    context 'valid param input' do
+    context 'invalid param input' do
```
Comment (hieunguyentrung) [inline spec/controllers/sessions_controller_spec.rb]: `invalid params input `

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -1,9 +1,39 @@
 module SessionsHelper
-  def log_in(user)
+  def assign_user_to_session(user)
```
Comment (hieunguyentrung) [inline app/helpers/sessions_helper.rb]: `save_user_to_session`

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -1,9 +1,39 @@
...
+      find_and_assign_by_cookies
+      assign_user_to_session(@current_user) if @current_user
+    end
+  end
+
+  def find_and_assign_by_session
+    @current_user = User.find_by(id: session[:user_id])
+  end
+
+  def find_and_assign_by_cookies
```
Comment (hieunguyentrung) [inline app/helpers/sessions_helper.rb]: `find_and_assign_by_cookie`

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

**Review by hieunguyentrung — APPROVED** [review]

**Pull request merged**